internal/util/image/disk_cache.go

```diff
@@ -0,0 +1,144 @@
+	"iter"
+	"os"
+	"path/filepath"
+	"slices"
+	"time"
+
+	"github.com/containerd/containerd/archive"
+	"github.com/containers/image/v5/docker/reference"
+	"github.com/opencontainers/go-digest"
+	ocispecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	errorutil "github.com/operator-framework/operator-controller/internal/util/error"
+	fsutil "github.com/operator-framework/operator-controller/internal/util/fs"
+)
+
+const ConfigDirLabel = "operators.operatorframework.io.index.configs.v1"
+
+func CatalogCache(basePath string) Cache {
+	return &diskCache{
+		basePath: basePath,
+		filterFunc: func(ctx context.Context, srcRef reference.Named, image ocispecv1.Image) (archive.Filter, error) {
+			_, specIsCanonical := srcRef.(reference.Canonical)
+			dirToUnpack, ok := image.Config.Labels[ConfigDirLabel]
+			if !ok {
+				// If the spec is a tagged ref, retries could end up resolving a new digest, where the label
+				// might show up. If the spec is canonical, no amount of retries will make the label appear.
+				// Therefore, we treat the error as terminal if the reference from the spec is canonical.
+				return nil, errorutil.WrapTerminal(fmt.Errorf("catalog image is missing the required label %q", ConfigDirLabel), specIsCanonical)
+			}
+
+			return allFilters(
+				onlyPath(dirToUnpack),
+				forceOwnershipRWX(),
+			), nil
+		},
+	}
+}
+
+func BundleCache(basePath string) Cache {
+	return &diskCache{
+		basePath: basePath,
+		filterFunc: func(_ context.Context, _ reference.Named, _ ocispecv1.Image) (archive.Filter, error) {
+			return forceOwnershipRWX(), nil
+		},
+	}
+}
+
+type diskCache struct {
+	basePath   string
+	filterFunc func(context.Context, reference.Named, ocispecv1.Image) (archive.Filter, error)
+}
+
+func (a *diskCache) Fetch(ctx context.Context, id string, canonicalRef reference.Canonical) (fs.FS, time.Time, error) {
+	l := log.FromContext(ctx)
+	unpackPath := a.unpackPath(id, canonicalRef.Digest())
+	modTime, err := fsutil.GetDirectoryModTime(unpackPath)
+	switch {
+	case errors.Is(err, os.ErrNotExist):
+		return nil, time.Time{}, nil
+	case errors.Is(err, fsutil.ErrNotDirectory):
+		l.Info("unpack path is not a directory; attempting to delete", "path", unpackPath)
+		return nil, time.Time{}, fsutil.DeleteReadOnlyRecursive(unpackPath)
+	case err != nil:
+		return nil, time.Time{}, fmt.Errorf("error checking image content already unpacked: %w", err)
+	}
+	l.Info("image already unpacked")
+	return os.DirFS(a.unpackPath(id, canonicalRef.Digest())), modTime, nil
+}
+
+func (a *diskCache) idPath(id string) string {
+	return filepath.Join(a.basePath, id)
+}
+
+func (a *diskCache) unpackPath(id string, digest digest.Digest) string {
+	return filepath.Join(a.idPath(id), digest.String())
+}
+
+func (a *diskCache) Store(ctx context.Context, id string, srcRef reference.Named, canonicalRef reference.Canonical, imgCfg ocispecv1.Image, layers iter.Seq[LayerData]) (fs.FS, time.Time, error) {
```

Comment:
First use of a go1.23 iterator!

I implemented this function this way in order to eliminate the cache's knowledge of the puller's internals. The puller provides the `layers` iterator to the cache. And the cache simply needs to iterate the layers.